core/trino-main/src/main/java/io/trino/connector/WorkerDynamicCatalogManager.java

```diff
@@ -91,29 +106,63 @@ public void ensureCatalogsLoaded(Session session, List<CatalogProperties> expect
             return;
         }
 
-        catalogsUpdateLock.lock();
+        catalogLoadingLock.lock();
         try {
             if (stopped) {
                 return;
             }
 
-            for (CatalogProperties catalog : getMissingCatalogs(expectedCatalogs)) {
-                checkArgument(!catalog.getCatalogHandle().equals(GlobalSystemConnector.CATALOG_HANDLE), "Global system catalog not registered");
-                CatalogConnector newCatalog = catalogFactory.createCatalog(catalog);
-                catalogs.put(catalog.getCatalogHandle(), newCatalog);
-                log.debug("Added catalog: " + catalog.getCatalogHandle());
+            List<CatalogProperties> missingCatalogs = getMissingCatalogs(expectedCatalogs);
+            missingCatalogs.forEach(catalog -> checkArgument(!catalog.getCatalogHandle().equals(GlobalSystemConnector.CATALOG_HANDLE), "Global system catalog not registered"));
+            List<ListenableFuture<Void>> loadedCatalogs = Futures.inCompletionOrder(
+                    missingCatalogs.stream()
+                            .map(catalog ->
+                                    Futures.submit(() -> {
+                                        catalogs.computeIfAbsent(catalog.getCatalogHandle(), ignore -> {
+                                            CatalogConnector newCatalog = catalogFactory.createCatalog(catalog);
+                                            log.debug("Added catalog: " + catalog.getCatalogHandle());
+                                            return newCatalog;
+                                        });
+                                    }, executor))
+                            .collect(toImmutableList()));
+
+            Deque<Throwable> catalogLoadingExceptions = new LinkedList<>();
+            for (ListenableFuture<Void> loadedCatalog : loadedCatalogs) {
+                try {
+                    // Timeout much longer than is expected to load a catalog
+                    // arbitrarily 3x TaskUpdateRequest timeout
+                    loadedCatalog.get(30, TimeUnit.SECONDS);
```

Comment:
A timeout is necessary here and introduces unwanted complexity. Creating the catalog should either succeed or fail -- there's nothing that should get stuck indefinitely in that process. If there is, there are larger problems and the server is already in a broken state (or will get into one).